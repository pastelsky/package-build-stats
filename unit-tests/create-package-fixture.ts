import fs from 'fs/promises'
import mkdirp from 'mkdirp'
// @ts-ignore
import PackageJson from '@npmcli/package-json'
import path from 'path'

function makeid(length: number) {
  let result = ''
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  const charactersLength = characters.length
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength))
  }
  return result
}

async function createPackageFile(
  packagePath: string,
  relativePath: string,
  content: string,
  isMain: boolean = false
) {
  const fullPath = path.join(packagePath, relativePath)
  await fs.writeFile(fullPath, content, 'utf8')
  if (isMain) {
    const pkgJson = await PackageJson.load(packagePath)
    pkgJson.update({
      main: relativePath,
    })
    await pkgJson.save()
  }
}

type PackageDescriptor = {
  name: string
  version: string
  path: string
}

async function addDependency(
  packagePath: string,
  name: string,
  linkImport: boolean = true
) {
  const dependencyPath = path.join(packagePath, 'node_modules')
  const pkgJson = await PackageJson.load(packagePath)
  if (linkImport) {
    const pkg = require(path.join(packagePath, 'package.json'))
    console.log('link import')
    const mainFile = pkg.main || pkg.module
    const mainFilePath = path.join(packagePath, mainFile)

    const mainFileContents = await fs.readFile(path.join(packagePath, mainFile))
    await fs.writeFile(
      mainFilePath,
      `import '${name}';\n ${mainFileContents}`,
      'utf8'
    )
  }
  const result = await createPackage(dependencyPath, name, { name: name })
  pkgJson.update({
    dependencies: {
      [name]: `*`,
    },
  })
  await pkgJson.save()
  return result
}

export async function createPackage(
  where: string,
  name: string,
  packageJSON: Record<string, any>
) {
  const packagePath = path.resolve(path.join(where, name))
  await mkdirp(packagePath)

  await fs.writeFile(
    path.join(packagePath, 'package.json'),
    JSON.stringify(packageJSON, null, 2)
  )

  return {
    get: () => ({
      name: packageJSON.name,
      version: packageJSON.version,
      path: path,
    }),
    addDependency: (name: string, linkImport: boolean = true) =>
      addDependency(packagePath, name, linkImport),
    addFile: (relativePath: string, content: string, isMain: boolean = false) =>
      createPackageFile(packagePath, relativePath, content, isMain),
  }
}

export function generateContent(sizeBytes: number) {
  const content = makeid(sizeBytes)
  return `
    console.log('${makeid(Math.max(0, sizeBytes - 15))}')
  `
}
