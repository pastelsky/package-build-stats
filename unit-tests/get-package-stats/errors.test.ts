import { createPackage, generateContent } from '../create-package-fixture'
import {
  getPackageStats,
  EntryPointError,
} from '../../packages/package-build-stats'
import path from 'path'
import semver from 'semver'
import Chance from 'chance'
import {
  BuildError,
  InstallError,
  PackageNotFoundError,
} from 'package-build-stats'

jest.mock('shortid', () => {
  return {
    generate: jest.fn().mockImplementation(() => 'test'),
  }
})

const withoutEntryPoint = async (
  additionalFields: Record<string, any> = {},
  contentSize: number = 100
) => {
  const packageRoot = path.resolve(path.join(__dirname, '..', 'staging'))
  const packagePath = path.join(packageRoot, 'without-entry-point')
  let pack = await createPackage(packageRoot, 'without-entry-point', {
    name: 'without-entry-point',
    ...additionalFields,
  })
  await pack.addFile('non-index.js', generateContent(contentSize), false)
  return packagePath
}

const withInvalidSyntax = async (
  additionalFields: Record<string, any> = {},
  contentSize: number = 100
) => {
  const packageRoot = path.resolve(path.join(__dirname, '..', 'staging'))
  const packagePath = path.join(packageRoot, 'with-invalid-syntax')
  let pack = await createPackage(packageRoot, 'with-invalid-syntax', {
    name: 'with-invalid-syntax',
    ...additionalFields,
  })
  await pack.addFile('index.js', `())${generateContent(contentSize)}`, false)
  return packagePath
}

const getError = async (promise: Promise<any>): Promise<any> => {
  try {
    await promise
  } catch (error) {
    return error
  }
}

describe('error tests', () => {
  let chance: Chance.Chance

  beforeAll(() => {
    chance = new Chance()
  })

  it.only('throws if entry point is not defined', async () => {
    const packagePath = await withoutEntryPoint()
    let statsPromise = getPackageStats(packagePath, {
      debug: true,
      client: 'pnpm',
    })
    await expect(statsPromise).rejects.toThrow(EntryPointError)
    const error = await getError(statsPromise)
    expect(error.originalError).toMatchSnapshot()
  })

  describe('throws if package is not available', () => {
    test.each(['npm', 'yarn', 'pnpm'])('throws for %p', async client => {
      let statsPromise = getPackageStats('some-path-not-existing', {
        debug: true,
        // @ts-ignore -- client is a string enum
        client,
      })
      await expect(statsPromise).rejects.toThrow(PackageNotFoundError)
      const error = await getError(statsPromise)
      expect(
        error.originalError
          .split('\n')
          .filter((line: string) => !line.includes('.log'))
          .join('\n')
      ).toMatchSnapshot()
    })
  })

  it.only('throws on invalid syntax', async () => {
    const packagePath = await withInvalidSyntax()
    let statsPromise = getPackageStats(packagePath, {
      debug: true,
      client: 'pnpm',
    })
    await expect(statsPromise).rejects.toThrow(BuildError)
    const error = await getError(statsPromise)
    expect(error.originalError).toMatchSnapshot()
  })
})
