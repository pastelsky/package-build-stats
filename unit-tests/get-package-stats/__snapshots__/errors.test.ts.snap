// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error tests throws if entry point is not defined 1`] = `
Array [
  Object {
    "codeFrames": Array [
      Object {
        "code": "const p = require('without-entry-point'); console.log(p)",
        "codeHighlights": Array [
          Object {
            "end": Object {
              "column": 39,
              "line": 1,
            },
            "start": Object {
              "column": 19,
              "line": 1,
            },
          },
        ],
        "filePath": "/private/tmp/tmp-build/packages/build-without-entry-point-tes/index.js",
      },
    ],
    "message": "Failed to resolve 'without-entry-point' from '/private/tmp/tmp-build/packages/build-without-entry-point-tes/index.js'",
    "origin": "@parcel/core",
  },
]
`;

exports[`error tests throws if package is not available throws for "npm" 1`] = `
"npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/some-path-not-existing - Not found
npm ERR! 404 
npm ERR! 404  'some-path-not-existing@*' is not in this registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
{
  \\"error\\": {
    \\"code\\": \\"E404\\",
    \\"summary\\": \\"Not Found - GET https://registry.npmjs.org/some-path-not-existing - Not found\\",
    \\"detail\\": \\"\\\\n 'some-path-not-existing@*' is not in this registry.\\\\nYou should bug the author to publish it (or use the name yourself!)\\\\n\\\\nNote that you can also install from a\\\\ntarball, folder, http url, or git url.\\"
  }
}

npm ERR! A complete log of this run can be found in:
"
`;

exports[`error tests throws if package is not available throws for "pnpm" 1`] = `
"[41m[30mâ€‰ERR_PNPM_FETCH_404â€‰[39m[49m [31mGET https://registry.npmjs.org/some-path-not-existing: Not Found - 404[39m

some-path-not-existing is not in the npm registry, or you have no permission to fetch it.

An authorization header was used: Bearer npm_[hidden]
"
`;

exports[`error tests throws if package is not available throws for "yarn" 1`] = `
"[91mâž¤[39m ]8;;https://yarnpkg.com/advanced/error-codes#yn0027---cant_suggest_resolutionsYN0027]8;;: [38;5;173msome-path-not-existing[39m[38;5;37m@[39m[38;5;37munknown[39m can't be resolved to a satisfying range
[91mâž¤[39m ]8;;https://yarnpkg.com/advanced/error-codes#yn0035---network_errorYN0035]8;;: The remote server failed to provide the requested resource
[91mâž¤[39m ]8;;https://yarnpkg.com/advanced/error-codes#yn0035---network_errorYN0035]8;;:   [38;5;111mResponse Code[39m: ]8;;https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404404 (Not Found)]8;;
[91mâž¤[39m ]8;;https://yarnpkg.com/advanced/error-codes#yn0035---network_errorYN0035]8;;:   [38;5;111mRequest Method[39m: GET
[91mâž¤[39m ]8;;https://yarnpkg.com/advanced/error-codes#yn0035---network_errorYN0035]8;;:   [38;5;111mRequest URL[39m: [38;5;170mhttps://registry.yarnpkg.com/some-path-not-existing[39m

[91mâž¤[39m Errors happened when preparing the environment required to run this command.
"
`;

exports[`error tests throws on invalid syntax 1`] = `[BuildError: Parenthesized expression cannot be empty]`;
